compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2362,6 +2366,168 @@ impl<'a> Parser<'a> {
+                                                        guard_hi: guard.span.shrink_to_hi(),
+                                                        ident,
+                                                        expr: expr.clone(),
+                                                    });
+                                                }
+                                            }
+                                        }
+
+                                        // help: extract the expr into a `const VAL: _ = expr`
+                                        let ident = match self.field {
+                                            Some(field) => field.ident.as_str().to_uppercase(),
+                                            None => "VAL".to_owned(),
+                                        };
+                                        err.subdiagnostic(
+                                            &that.parser.dcx(),
+                                            UnexpectedExpressionInPatternConstSugg {
+                                                stmt_lo: line_lo,
+                                                ident_span: expr_span,
+                                                expr,
+                                                ident,
+                                                indentation,
+                                            },
+                                        );
+
+                                        // help: wrap the expr in a `const { expr }`
+                                        // FIXME(inline_const): once stabilized, remove this check and remove the `(requires #[feature(inline_const])` note from the message
+                                        if that.parser.psess.unstable_features.is_nightly_build() {
+                                            err.subdiagnostic(
+                                                &that.parser.dcx(),
+                                                UnexpectedExpressionInPatternInlineConstSugg {
+                                                    start_span: expr_span.shrink_to_lo(),
+                                                    end_span: expr_span.shrink_to_hi(),
+                                                },
+                                            );
+                                        }
+                                    }
+                                },
+                            );
+                    }; // end of `emit_now` closure, we're back in `visit_pat`
+
+                    match &p.kind {
+                        // Base expression
+                        PatKind::Err(_) => emit_now(self, p.span, p.span),
+                        // Sub-patterns
+                        PatKind::Box(subpat) | PatKind::Ref(subpat, _)
+                            if matches!(subpat.kind, PatKind::Err(_)) =>
+                        {
+                            emit_now(self, subpat.span, p.span)
+                        }
```

Comment:
This should be a loop (as to handle `&mut &mut PatKind::Err`), but I'm not sure if it isn't better to just returns `PatKind::Err` instead of `PatKind::Ref(PatKind::Err, _)` when parsing *ReferencePattern*s if the subpat isn't valid, at the cost of loosing information.

Same with `PatKind::Range`, when one of its bounds is an `ExprKind::Err`.